velox/common/compression/Compression.h

```diff
@@ -43,7 +46,233 @@ std::string compressionKindToString(CompressionKind kind);
 
 CompressionKind stringToCompressionKind(const std::string& kind);
 
-constexpr uint64_t DEFAULT_COMPRESSION_BLOCK_SIZE = 256 * 1024;
+static constexpr uint64_t kDefaultCompressionBlockSize = 256 * 1024;
+
+static constexpr int32_t kDefaultCompressionLevel =
+    std::numeric_limits<int32_t>::min();
+
+class StreamingCompressor;
+class StreamingDecompressor;
+
+struct CodecOptions {
+  int32_t compressionLevel;
+
+  CodecOptions(int32_t compressionLevel = kDefaultCompressionLevel)
+      : compressionLevel(compressionLevel) {}
+
+  virtual ~CodecOptions() = default;
+};
+
+/// Codec interface for compression and decompression.
+/// The Codec class provides a common interface for various compression
+/// algorithms to support one-shot compression and decompression.
+///
+/// For codecs that support streaming compression and decompression, the
+/// makeStreamingCompressor() and makeStreamingDecompressor() functions can be
+/// used to create streaming compressor and decompressor instances.
+class Codec {
+ public:
+  virtual ~Codec() = default;
+
+  // Create a kind for the given compression algorithm with CodecOptions.
+  static Expected<std::unique_ptr<Codec>> create(
+      CompressionKind kind,
+      const CodecOptions& codecOptions = CodecOptions{});
+
+  // Create a kind for the given compression algorithm.
+  static Expected<std::unique_ptr<Codec>> create(
+      CompressionKind kind,
+      int32_t compressionLevel);
+
+  // Return true if support for indicated kind has been enabled.
+  static bool isAvailable(CompressionKind kind);
+
+  /// Return true if indicated kind supports extracting uncompressed length
+  /// from compressed data.
+  static bool supportsGetUncompressedLength(CompressionKind kind);
+
+  /// Return true if indicated kind supports one-shot compression with fixed
+  /// compressed length.
+  static bool supportsCompressFixedLength(CompressionKind kind);
+
+  /// Return the smallest supported compression level.
+  /// If the codec doesn't support compression level,
+  /// `kUseDefaultCompressionLevel` will be returned.
+  virtual int32_t minCompressionLevel() const = 0;
+
+  /// Return the largest supported compression level.
+  /// If the codec doesn't support compression level,
+  /// `kUseDefaultCompressionLevel` will be returned.
+  virtual int32_t maxCompressionLevel() const = 0;
+
+  /// Return the default compression level.
+  /// If the codec doesn't support compression level,
+  /// `kUseDefaultCompressionLevel` will be returned.
+  virtual int32_t defaultCompressionLevel() const = 0;
+
+  /// Performs one-shot compression.
+  /// `outputLength` must first have been computed using maxCompressedLength().
+  /// The actual compressed length will be written to actualOutputLength.
+  /// Note: One-shot compression is not always compatible with streaming
+  /// decompression. Depending on the codec (e.g. LZ4), different formats may
+  /// be used.
```

Comment:
Let's add to this comment what happens if the output buffer is null or if the length is 0.

Reply:
The ubsan error is reported from `decompress`.

Normally, the compression output won't be zero-sized. But for decompression, the output buffer can be null if the decompressed size is zero. Updated the comments for both functions. Please help to review again. Thanks!

Reply:
@marin-ma the build is still broken as we're still calling the lz4 function with a nullptr buffer. We need to add a check to prevent calling it in such cases, or the ubsan build won't go through.

Reply:
@pedroerp updated the code to prevent passing nullptr to `output`. Could you take an another look? Thanks!